pytest >= 8.2 is not yet supported (not only tests fail)

See https://github.com/pytest-dev/pytest-rerunfailures/actions/runs/8409152908/job/23026172136

This ~~can be a temporary problem or a~~ is a hint that we have to change something here to make it compatible again.

With pytest==8.2.2 and pytest-xdist==3.6.1, and pytest-rerunfailures==14.0, all tests that undergo a rerun result in the following AssertionError:

```
    def runtest(self) -> None:
        from _pytest.debugging import maybe_wrap_pytest_function_for_tracing
    
        testcase = self.instance
>       assert testcase is not None
E       AssertionError
```

It works fine for me on pytest==8.2.1 and pytest-cdist==3.5.0 and pytest-rerunfailures==14.0.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -4,7 +4,8 @@ Changelog
 14.1 (unreleased)
 -----------------
 
-- Nothing changed yet.
+- Fix compatibility with pytest 8.2.
+  (`#267 <https://github.com/pytest-dev/pytest-rerunfailures/issues/267>`_)
 
 
 14.0 (2024-03-13)

--- a/src/pytest_rerunfailures.py
+++ b/src/pytest_rerunfailures.py
@@ -468,6 +468,9 @@ class ClientStatusDB(SocketDB):
         return int(self._sock_recv(self.sock))
 
 
+suspended_finalizers = {}
+
+
 def pytest_runtest_teardown(item, nextitem):
     reruns = get_reruns_count(item)
     if reruns is None:
@@ -490,13 +493,20 @@ def pytest_runtest_teardown(item, nextitem):
         and any(_test_failed_statuses.values())
         and not any(item._terminal_errors.values())
     ):
-        # clean cashed results from any level of setups
+        # clean cached results from any level of setups
         _remove_cached_results_from_failed_fixtures(item)
 
         if item in item.session._setupstate.stack:
             for key in list(item.session._setupstate.stack.keys()):
                 if key != item:
+                    # only the first finalizer contains the correct teardowns
+                    if key not in suspended_finalizers:
+                        suspended_finalizers[key] = item.session._setupstate.stack[key]
                     del item.session._setupstate.stack[key]
+    else:
+        # restore suspended finalizers
+        item.session._setupstate.stack.update(suspended_finalizers)
+        suspended_finalizers.clear()
 
 
 @pytest.hookimpl(hookwrapper=True)

